network/src/router.rs

```diff
@@ -146,52 +219,143 @@ impl<P: ProvideRuntimeApi + Send + Sync + 'static> Router<P>
+
+				debug!(target: "consensus", "Circulating messages from {:?} to {:?} at {}",
+					source, target, self.parent_hash);
+
+				// this is the ingress from source to target, with given messages.
+				let target_incoming = incoming_message_topic(self.parent_hash, target);
+				let ingress_for: IngressPairRef = (source, &group_messages[..]);
+
+				self.network.gossip_message(target_incoming, ingress_for.encode(), false);
 			}
 		}
 	}
 
 	fn create_work<D>(&self, candidate_hash: Hash, producer: ParachainWork<D>)
-		-> impl Future<Item=(),Error=()>
+		-> impl Future<Item=(),Error=()> + Send + 'static
 		where
-		D: Future<Item=BlockData,Error=io::Error> + Send + 'static,
+		D: Future<Item=(BlockData, Incoming),Error=io::Error> + Send + 'static,
 	{
 		let table = self.table.clone();
 		let network = self.network.clone();
 		let knowledge = self.knowledge.clone();
 		let attestation_topic = self.attestation_topic.clone();
 
 		producer.prime(self.api.clone())
-			.map(move |produced| {
+			.map(move |validated| {
 				// store the data before broadcasting statements, so other peers can fetch.
 				knowledge.lock().note_candidate(
 					candidate_hash,
-					Some(produced.block_data),
-					produced.extrinsic,
+					Some(validated.block_data().clone()),
+					validated.extrinsic().cloned(),
 				);
 
 				// propagate the statement.
 				// consider something more targeted than gossip in the future.
-				let signed = table.sign_and_import(produced.validity);
-				network.gossip_consensus_message(attestation_topic, signed.encode(), false);
+				let signed = table.import_validated(validated);
+				network.gossip_message(attestation_topic, signed.encode(), false);
 			})
 			.map_err(|e| debug!(target: "p_net", "Failed to produce statements: {:?}", e))
 	}
+
+}
+
+impl<P: ProvideRuntimeApi, N, T> Router<P, N, T> where
+	P::Api: ParachainHost<Block>,
+	N: NetworkService,
+	T: Executor,
+{
+	fn do_fetch_incoming(&self, parachain: ParaId) -> IncomingReceiver {
+		use polkadot_primitives::BlockId;
+		let (tx, rx) = {
+			let mut fetching = self.fetch_incoming.lock();
+			match fetching.entry(parachain) {
+				Entry::Occupied(entry) => return entry.get().clone(),
+				Entry::Vacant(entry) => {
+					// has not been requested yet.
+					let (tx, rx) = oneshot::channel();
+					let rx = IncomingReceiver { inner: rx.shared() };
+					entry.insert(rx.clone());
+
+					(tx, rx)
+				}
+			}
+		};
+
+		let parent_hash = self.parent_hash;
+		let topic = incoming_message_topic(parent_hash, parachain);
+		let gossip_messages = self.network.gossip_messages_for(topic)
+			.map_err(|()| panic!("unbounded receivers do not throw errors; qed"))
+			.filter_map(|msg| IngressPair::decode(&mut msg.as_slice()));
+
+		let canon_roots = self.api.runtime_api().ingress(&BlockId::hash(parent_hash), parachain)
+			.map_err(|e| format!("Cannot fetch ingress for parachain {:?} at {:?}: {:?}",
+				parachain, parent_hash, e)
+			);
+
+		// TODO: select with `Exit` here.
```

Comment:
This doesn't seem to be hard, should we just do that or something prevents us for that?